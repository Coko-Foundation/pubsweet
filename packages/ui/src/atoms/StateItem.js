import PropTypes from 'prop-types'
import styled from 'styled-components'
import React from 'react'

const darkGrey = '#404040'
const lightGrey = '#b3b3b3'

const disabled = `
  color: ${lightGrey};
  cursor: default;

  &:hover {
    color: ${lightGrey};
  }
`

const Root = styled.span`
  cursor: pointer;
  font-family: var(--font-interface);
  font-size: 16px;
  font-style: italic;
  padding: 0;

  &:focus {
    outline: none;
  }

  &:hover {
    color: ${darkGrey};
    transition: 0.25s ease-in-out 0s;
  }

  ${props => (props.disabled ? disabled : '')};
`

const StateItem = ({ update, disabled, values, index }) => {
  const callUpdateWithNextIndex = () => {
    if (disabled) return
    const nextIndex = arrayShift(values, index)
    update(values[index], nextIndex)
  }

  const arrayShift = (array, i) => (i === array.length - 1 ? 0 : i + 1)

  return (
    <Root
      disabled={disabled}
      onClick={callUpdateWithNextIndex}
      onKeyPress={callUpdateWithNextIndex}
      role="button"
      tabIndex="0"
    >
      {values[index]}
    </Root>
  )
}

StateItem.propTypes = {
  disabled: PropTypes.bool,
  index: PropTypes.number.isRequired,
  update: PropTypes.func.isRequired,
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
}

StateItem.defaultProps = {
  disabled: false,
}

export default StateItem
