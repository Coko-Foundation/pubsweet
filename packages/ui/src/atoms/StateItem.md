An interactive element which renders a text element from a given index of a provided array and upon click runs a provided update function, passing it the current text element and the index of the next item in the array, or 0 if the current item is the last.

If the update function is used to update the index, the component will cycle through the array on click.

```js
const initialState = {
  values: ['To Clean', 'Cleaning', 'Cleaned'],
  index: 0,
}

const update = (currentValue, nextIndex) => {
  setState({ index: nextIndex })
}
;<StateItem values={state.values} index={state.index} update={update} />
```

If the component is passed the disabled prop, the update function is not run:

```js
const initialState = {
  values: ['To Clean', 'Cleaning', 'Cleaned'],
  index: 0,
}

const update = (currentValue, nextIndex) => {
  setState({ index: nextIndex })
}
;<StateItem
  values={state.values}
  index={state.index}
  disabled={true}
  update={update}
/>
```
