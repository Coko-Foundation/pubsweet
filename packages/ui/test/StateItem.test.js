import React from 'react'
import { clone } from 'lodash'
import { shallow, render } from 'enzyme'
import 'jest-styled-components'
import renderer from 'react-test-renderer'

import StateItem from '../src/atoms/StateItem'

const myMock = jest.fn()
const props = {
  values: ['To Clean', 'Cleaning', 'Cleaned'],
  update: myMock,
  index: 1,
}

const wrapper = shallow(<StateItem {...props} />)
const wrapperRendered = render(<StateItem {...props} />)

describe('StateItem', () => {
  test('is rendered correctly', () => {
    const tree = renderer.create(<StateItem {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with default props cursor should be pointer', () => {
    const tree = renderer.create(<StateItem {...props} />).toJSON()
    expect(tree).toHaveStyleRule('cursor', 'pointer')
  })

  test('with disabled prop cursor should be default', () => {
    const newProps = clone(props)
    newProps.disabled = true
    const tree = renderer.create(<StateItem {...newProps} />).toJSON()
    expect(tree).toHaveStyleRule('cursor', 'default')
  })

  test('should render the value Cleaning', () => {
    expect(wrapperRendered.text()).toEqual(props.values[props.index])
  })

  test('update method should be triggered upon click', () => {
    wrapper.simulate('click')
    expect(props.update).toHaveBeenCalled()
  })
})
