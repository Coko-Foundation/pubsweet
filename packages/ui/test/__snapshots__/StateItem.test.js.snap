// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateItem is rendered correctly 1`] = `
.c0 {
  cursor: pointer;
  font-family: var(--font-interface);
  font-size: 16px;
  font-style: italic;
  padding: 0;
}

.c0:focus {
  outline: none;
}

.c0:hover {
  color: #404040;
  -webkit-transition: 0.25s ease-in-out 0s;
  transition: 0.25s ease-in-out 0s;
}

<span
  className="c0"
  disabled={false}
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  Cleaning
</span>
`;
